### PR TITLE
Feat/balsamic store completed

### PR DESCRIPTION
**Pull request by patrikgrenfeldt**
This PR adds a command to store multiple analyses at once

N.B. Re-try from: https://github.com/Clinical-Genomics/cg/pull/632

**How to prepare for test**:
- [x] ssh to hasta (depending on type of change)
- [x] install on stage:
`bash servers/resources/hasta.scilifelab.se/update-cg-stage.sh [THIS-BRANCH-NAME]`

**How to test**:
- [x] login to hasta
- [x] do cg workflow balsamic store completed

**Expected test outcome**:
- [x] check that it stores several cases at once
- [x] Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by MR
- [x] tests executed by @patrikgrenfeldt 
- [x] "Merge and deploy" approved by @patrikgrenfeldt 
Thanks for filling in who performed the code review and the test!

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: cg/cli/workflow/balsamic/store.py [hunk 1]
```diff
@@ -104,6 +134,35 @@ def generate_deliverables_file(context, dry, config_path, case_id):
...
+        except StoreError as error:
+            LOG.error("Analysis storage failed: %s", error.message)
+            exit_code = FAIL
+        except FileNotFoundError as error:
+            LOG.error("Analysis storage failed, missing file: %s", error.args[0])
+            exit_code = FAIL
+
+    click.echo(click.style(f"Done storing cases. Exit code: {exit_code}", fg="blue"))
+
+    click.echo(click.style(f"Done storing cases. Exit code: {exit_code}", fg="blue"))
```
Comment (patrikgrenfeldt) [inline cg/cli/workflow/balsamic/store.py:162]:
```suggestion
```

File: cg/cli/workflow/balsamic/store.py [hunk 2]
```diff
@@ -23,77 +25,105 @@
...
+
+    for case in _store.families():
+        try:
+            click.echo(click.style(f"storing case: {case}", fg="blue"))
+            context.invoke(analysis, case_id=case.internal_id)
+        except CgError as error:
+            LOG.error(error.message)
+            exit_code = 1
+
+    sys.exit(exit_code)
```
Comment (patrikgrenfeldt) [inline cg/cli/workflow/balsamic/store.py:102]:
```suggestion
```

File: cg/cli/workflow/balsamic/store.py [hunk 3]
```diff
@@ -23,77 +25,105 @@
 def store(context):
     """Store results from MIP in housekeeper."""
```
Comment (moonso) [inline cg/cli/workflow/balsamic/store.py:26]:
```suggestion
    """Store results from Balsamic in housekeeper."""
```

File: cg/cli/workflow/balsamic/store.py [hunk 4]
```diff
@@ -23,77 +25,105 @@
...
 
     if not case_obj:
-        click.echo(click.style(f"Case {case_id} not found", fg="yellow"))
+        raise CgError(f"Case {case_id} not found")
 
     if not config_path:
-        config_path = Path.joinpath(root_dir, case_id, case_id + ".json")
+        config_path = analysis_api.get_config_path(root_dir, case_id)
+        if not config_path.is_file():
+            raise FileNotFoundError(f"Missing the sample-config file for {case_id}: {config_path}")
```
Comment (patrikgrenfeldt) [inline cg/cli/workflow/balsamic/store.py:124]:
```suggestion
```

File: cg/cli/workflow/balsamic/store.py [hunk 5]
```diff
@@ -23,77 +25,105 @@
 def store(context):
     """Store results from MIP in housekeeper."""
     context.obj["db"] = Store(context.obj["database"])
-    context.obj["tb_api"] = tb.TrailblazerAPI(context.obj)
     context.obj["hk_api"] = hk.HousekeeperAPI(context.obj)
+    context.obj["analysis_api"] = AnalysisAPI(
```
Comment (moonso) [inline cg/cli/workflow/balsamic/store.py:29]: In this not instanciated earlier in the balsamic base command?

File: cg/cli/workflow/balsamic/store.py [hunk 6]
```diff
@@ -23,77 +25,105 @@
...
+
+@store.command("all-cases")
+@click.pass_context
+def all_cases(context):
+
+    _store = context.obj["db"]
+
+    for case in _store.families():
+        try:
+            click.echo(click.style(f"storing case: {case}", fg="blue"))
```
Comment (moonso) [inline cg/cli/workflow/balsamic/store.py:96]:
```suggestion
            LOG.info("storing case: %s", case)
```

File: cg/cli/workflow/balsamic/store.py [hunk 7]
```diff
@@ -104,6 +134,35 @@ def generate_deliverables_file(context, dry, config_path, case_id):
...
+@click.pass_context
+def completed(context):
+    """Store all completed analyses."""
+    _store = context.obj["db"]
+
+    exit_code = SUCCESS
+    for case in _store.cases_to_balsamic_analyze(limit=None):
+        click.echo(click.style(f"Storing case: {case}", fg="blue"))
+        try:
+            exit_code = context.invoke(analysis, case_id=case.internal_id) and exit_code
```
Comment (patrikgrenfeldt) [inline cg/cli/workflow/balsamic/store.py:152]:
```suggestion
            exit_code = context.invoke(analysis, case_id=case.internal_id) or exit_code
```

**Comment by moonso** [comment]
Nice work both of you! So I leave it as is, I guess we don't need another review if you already looked at the code together

**Comment by patrikgrenfeldt** [comment]
Thanks for the review @Mropat 😄

File: cg/cli/workflow/balsamic/store.py (hunk 5 above)
Comment (patrikgrenfeldt) [reply cg/cli/workflow/balsamic/store.py]: utilised now

File: cg/cli/workflow/balsamic/store.py (hunk 6 above)
Comment (patrikgrenfeldt) [reply cg/cli/workflow/balsamic/store.py]: logging instead of echoing now

**Comment by patrikgrenfeldt** [comment]
Deployed on Hasta: 
image

**Pull request merged**